### ISSUE TITLE
parser registry

### ISSUE THREAD

**Issue by anntzer** [evanunderscore/defopt#69]
defopt parsers are a bit annoying to reuse across modules: quite often I introduce a "pseudo-type" solely for the purposes of making a defopt annotation (e.g., `comma_separated_int_list` with a parser that's `def parse_comma_separated_int_list(s): return [*map(int, s.split(","))] if s else []` (*)), but then to reuse the parser in another module I need to import the type annotation and the parser (well, I can cheat by making the type and the parser the same object...) and also pass it explicitly to the `defopt.main` call.

((*) not just using List[int] because I want List[comma_separated_int_list], in fact...)

It may be useful if it was possible to register parsers with defopt, e.g.
`custom_type.py`
```python
class foo: ...

@defopt.register_parser(foo)
def parse_foo(s): ...
```
and then
`calling_module.py`
```python
from custom_type import foo

def main(arg: foo): ...

defopt.run(main)  # no need to pass foo as parser.
```

Alternatively, to at least handle the case where the type annotation and the parser are the same object (i.e., an object of class foo can be constructed just from `foo(s: str)`), when defopt sees an unknown type `foo`, it could e.g. check whether the signature of `foo` consists of a single annotated-as-str parameter, and if so, assume that `foo` is indeed its own parser.
Thoughts?

**Comment by evanunderscore** [comment]
defopt actually used to *require* you to register parsers exactly like this (there was no option to pass arguments to `run`), but I changed it in 1.0.0 because it felt over-engineered for my use cases and the way it works now felt simpler. However at this point you're certainly using the module more than I am, so I'm happy to defer to your judgement.

On assuming that `foo(s: str)` is its own parser: yes, that sounds like a sensible thing to do by default. I'd be happy to accept that change if you'd like to implement it.

On your particular use case of defining `comma_separated_int_list`: from a purist standpoint, I never wanted people to have to do things like that in order to manipulate the generated command line. In this particular case, with a small fix, this becomes possible:

```
diff --git a/defopt.py b/defopt.py
index 80c1445..ec0cbe2 100644
--- a/defopt.py
+++ b/defopt.py
@@ -335,7 +335,7 @@ def _get_type_from_hint(hint):
         list, collections_abc.Iterable, collections_abc.Sequence,  # Py>=3.7
     ]
     if ti.get_origin(hint) in container_types:
-        [type_] = ti.get_args(hint)
+        [type_] = ti.get_args(hint, evaluate=True)
         return _Type(type_, list)
     elif ti.is_union_type(hint):
         # For Union[type, NoneType], just use type.
diff --git a/examples/list_parser.py b/examples/list_parser.py
new file mode 100644
index 0000000..89063d8
--- /dev/null
+++ b/examples/list_parser.py
@@ -0,0 +1,17 @@
+from typing import List
+
+import defopt
+
+
+def example(list_of_lists_of_ints: List[List[int]]):
+    print(list_of_lists_of_ints)
+
+
+def int_list(string):
+    if not string:
+        return []
+    return [int(x) for x in string.split(',')]
+
+
+if __name__ == '__main__':
+    defopt.run(example, parsers={List[int]: int_list})
```

```
$ PYTHONPATH=. python examples/list_parser.py -l 1,2,3 4,5 6
[[1, 2, 3], [4, 5], [6]]
```

**Comment by anntzer** [comment]
> However at this point you're certainly using the module more than I am, so I'm happy to defer to your judgement.

I definitely use it a *lot*, and it's really useful.  Thanks again for writing it :)

> On assuming that foo(s: str) is its own parser.

I think that's a slightly less invasive change than explicitly having a parser registry and will probably start by implementing that.

> On your particular use case of defining comma_separated_int_list [...]

I agree my approach is a bit of a hack, and didn't think of adding a parser for List[int].  It looks like even the addition of `evaluate=True` is unnecessary, though?

**Comment by evanunderscore** [comment]
> Thanks again for writing it :)

No problem, thank you for your many contributions!

>  It looks like even the addition of `evaluate=True` is unnecessary, though?

It was necessary for me in 3.6. The docstring for `ti.get_args()` mentions the default value is `True` on 3.7, so if you're using that, that explains why you don't need it.

**Comment by anntzer** [comment]
I think this can be considered closed by #70 and #72.
